### PR TITLE
Re-upgrade term to 0.4(.3)

### LINKED ISSUE #2338: cargo commands are exiting prematurely (under Emacs `M-x shell`)
Update: This bug is occurring on Mac OS X in a particular (perhaps broken) terminal setup. Still I would like to know why cargo is not behaving properly.

I'm not sure if this is something wrong with my particular setup, or if something has recently changed in Rust or Cargo that would cause this, but I am finding that my cargo invocations (both valid commands and erroneous ones) are exiting prematurely; it seems like the main thread finishes and exits the process, without e.g. flushing buffered I/O. This makes for a frustrating user experience if you don't suspect it is happening.

Concrete examples (where the cases that I pipe the output of `cargo` to `cat` cause it to actually finish the desired command, AFAICT).

```
% pwd
/Users/fklock/Dev/Rust/tango
% ls
Cargo.lock	Cargo.toml~	LICENSE-MIT	src
Cargo.toml	LICENSE-APACHE	README.md	tests
% cargo --version
% cargo --version | cat
cargo 0.9.0-nightly (6ffd134 2016-01-26)
% cargo build
% echo $?
101
% cargo build | cat
   Compiling winapi-build v0.1.1
   Compiling libc v0.2.6
   Compiling kernel32-sys v0.2.1
   Compiling walkdir v0.1.5
   Compiling filetime v0.1.9
   Compiling tango v0.3.3 (file:///Users/fklock/Dev/Rust/tango)
% echo $?
0
% 
```

### PR DESCRIPTION
**Pull request by Stebalien**
This PR re-upgrades term to 0.4.3. Cargo now detects whether or not
color is supported immediately after creating the TerminfoTerminal.

Backstory:

Before v0.4, term used to return `Ok(true)` if something succeeded,
`Ok(false)` if the operation was unsupported, and `Err(io::Error)` if there
was an IO error. Now, it returns `Ok(())` if the operation succeeds and
`Err(term::Error)` if the operation fails. If the operation is unsupported,
it returns `Err(term::Error::NotSupported)`. This means that, if `op` is
unsupported, `try!(term.op())` will now return an error instead of silently
failing (well, return false but that's effectively silent).

Fixes #2338.

### CODE REVIEW

File: src/cargo/core/shell.rs [hunk 1]
```diff
@@ -149,21 +149,23 @@ impl Shell {
...
-            }
-            _ => {
-                // No color output.
-                Shell { terminal: NoColor(out), config: config }
-            }
-        }
+            Err(_) => Err(out)
+        }.unwrap_or_else(|out| Shell {
+            // If config.tty is set, we should be able to color output but can't. Write to stderr.
+            terminal: NoColor(if config.tty { Box::new(io::stderr()) } else { out }),
```
Comment (Stebalien) [inline src/cargo/core/shell.rs]: This is fishy. I'm currently just doing what cargo used to do but TTY being a terminal has no bearing on whether or not we can color. IMO, we should get rid of this check entirely. Why is it here anyways?

Comment (alexcrichton) [inline src/cargo/core/shell.rs]: I... have no idea what this is here nor what it is doing, it does indeed seem like it should be removed to me. Does it cause any tests to fail if removed?

**Comment by alexcrichton** [comment]
@pnkfelix could you try out Cargo with this patch and see if it works for you?

**Comment by pnkfelix** [comment]
@alexcrichton I will try it out

**Comment by pnkfelix** [comment]
I keep getting:

```
% cargo build --verbose
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading term v0.4.3
unable to get packages from source

Caused by:
  failed to unpack package `term v0.4.3`

Caused by:
  No such file or directory (os error 2)
```

And then subsequent build attempts likewise fail:

```
% cargo build --verbose
unable to get packages from source

Caused by:
  failed to unpack package `term v0.4.3`

Caused by:
  No such file or directory (os error 2)
```

(I've then tried forcing cargo to redo the download by deleting the `rm -rf ~/.multirust/toolchains/nightly-2016-01-26/cargo/registry`, which just causes `cargo build` to repeat of the top most command line transcript in this comment.)

**Comment by pnkfelix** [comment]
@alexcrichton okay, on a proper nightly (and doing the build itself in a non `emacs M-x shell` terminal, since the `cargo` from nightly won't work for the bootstrap in that procedure), the resulting `cargo` seems to work under the `emacs M-x shell` terminal.

**Comment by pnkfelix** [comment]
@alexcrichton but should I be concerned about the download failure for `term 0.4.3` when using the `2016-01-26` multirust toolchain? Super weird, but maybe some *other* bug has been fixed in the past two nights?

**Comment by Stebalien** [comment]
This is also causing the automated build to fail.

On February 1, 2016 5:52:16 AM EST, Felix S Klock II <notifications@github.com> wrote:
>@alexcrichton but should I be concerned about the download failure for
>`term 0.4.3` when using the `2016-01-26` multirust toolchain? Super
>weird, but maybe some *other* bug has been fixed in the past two
>nights?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rust-lang/cargo/pull/2344#issuecomment-177911089

-- 
Steven Allen

File: src/cargo/core/shell.rs (hunk 1 above)
Comment (Stebalien) [inline src/cargo/core/shell.rs]: It doesn't cause any additional test failures (34 tests fail on my machine regardless of what I do). I've pushed the change to this branch. If you'd like, I can break it out into a separate PR but I figured I'd fix it here as I had to rewrite this function anyways.

**Comment by Stebalien** [comment]
Sorry, I had to rebase to fix my email/name. My git name and email somehow got set to bar and baz...

**Comment by Stebalien** [comment]
> Sorry, I had to rebase to fix my email/name. My git name and email somehow got set to bar and baz...

Wait just a minute... @alexcrichton Why is your test case modifying MY git config? :(

**Comment by pnkfelix** [comment]
> Why is your test case modifying MY git config? :(

What the what?  Pointers please?

**Comment by Stebalien** [comment]
See #2345.
-- 
Steven Allen

**Comment by alexcrichton** [comment]
It looks like this recent publish may be suffering from https://github.com/rust-lang/cargo/issues/2326, can you try republishing with a newer Cargo?

**Comment by alexcrichton** [comment]
@pnkfelix thanks for the confirmation of it working!

I'm a little worried that we're just dropping errors on the floor somewhere, it would probably be good to track those down...

**Comment by Stebalien** [comment]
It should work now.

File: Cargo.toml [hunk 2]
```diff
@@ -37,7 +37,7 @@ regex = "0.1"
 rustc-serialize = "0.3"
 semver = "0.2.0"
 tar = "0.4"
-term = "0.2"
+term = "^0.4.4"
```
Comment (alexcrichton) [inline Cargo.toml]: Could this just be `0.4.4`? I believe the `^` is the default

Comment (Stebalien) [inline Cargo.toml]: Didn't realize that. Fixed.

**Comment by alexcrichton** [comment]
Thanks @Stebalien!

**Pull request merged**